internal/indexcoord/task.go

```diff
@@ -73,11 +81,159 @@ func (bt *BaseTask) Notify(err error) {
 	bt.done <- err
 }
 
-// IndexAddTask is used to record the information of the index tasks.
+// CreateIndexTask is used to create an index on field.
+type CreateIndexTask struct {
+	BaseTask
+	dataCoordClient  types.DataCoord
+	indexCoordClient *IndexCoord
+	req              *indexpb.CreateIndexRequest
+	indexID          UniqueID
+	createTs         uint64
+	idAllocator      *allocator.GlobalIDAllocator
+	tsoAllocator     *tso.GlobalTSOAllocator
+}
+
+// Ctx returns the context of the index task.
+func (cit *CreateIndexTask) Ctx() context.Context {
+	return cit.ctx
+}
+
+// ID returns the id of the index task.
+func (cit *CreateIndexTask) ID() UniqueID {
+	return cit.id
+}
+
+// SetID sets the id for index tasks.
+func (cit *CreateIndexTask) SetID(ID UniqueID) {
+	cit.BaseTask.setID(ID)
+}
+
+// Name returns the task name.
+func (cit *CreateIndexTask) Name() string {
+	return CreateIndexTaskName
+}
+
+// OnEnqueue assigns the indexBuildID to index task.
+func (cit *CreateIndexTask) OnEnqueue() error {
+	var err error
+	cit.indexID, err = cit.idAllocator.AllocOne()
+	if err != nil {
+		return err
+	}
+	cit.createTs, err = cit.tsoAllocator.AllocOne()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// PreExecute do nothing.
+func (cit *CreateIndexTask) PreExecute(ctx context.Context) error {
+	log.Info("IndexCoord CreateIndexTask PreExecute", zap.Int64("collectionID", cit.req.CollectionID),
+		zap.Int64("fieldID", cit.req.FieldID), zap.String("indexName", cit.req.IndexName))
+	return nil
+}
+
+// Execute adds the index task to meta table.
+func (cit *CreateIndexTask) Execute(ctx context.Context) error {
+	log.Info("IndexCoord CreateIndexTask Execute", zap.Int64("collectionID", cit.req.CollectionID),
+		zap.Int64("fieldID", cit.req.FieldID), zap.String("indexName", cit.req.IndexName))
+	hasIndex, indexID := cit.table.HasSameReq(cit.req)
+	if hasIndex {
+		cit.indexID = indexID
+	}
+	index := &model.Index{
+		CollectionID: cit.req.CollectionID,
+		FieldID:      cit.req.FieldID,
+		IndexID:      cit.indexID,
+		IndexName:    cit.req.IndexName,
+		TypeParams:   cit.req.TypeParams,
+		IndexParams:  cit.req.IndexParams,
+		CreateTime:   cit.createTs,
+	}
+	cit.table.SetIndex(index)
```

Comment:
do not use SetIndex.
first createSegmentIndex
then createIndex
finally noitfy builder to build index.